specutils/analysis/snr.py

```diff
@@ -0,0 +1,106 @@
+import numpy as np
+from ..utils import SpectralRegion
+
+__all__ = ['snr']
+
+
+def snr(spectrum, region=None, noise_region=None):
+    """
+    Calculate the mean S/N of the spectrum based on the flux and uncertainty in the spectrum. This will
+    be calculated over the regions, if they are specified.
+
+    Parameters
+    ----------
+    spectrum : ``specutils.spectra.spectrum1d.Spectrum1D``
+        The spectrum object overwhich the equivalent width will be calculated.
+
+    region: ``specutils.utils.SpectralRegion`` or list of ``specutils.utils.SpectralRegion``
+        Region within the spectrum to calculate the SNR.
+
+    noise_region: ``specutils.utils.SpectralRegion``
+        Region within the spectrum from which to calculate the noise (based on standard deviation
+        of the flux values)
+
+    Returns
+    -------
+    snr : float or list (based on region input)
+        Signal to noise ratio of the spectrum or within the regions
+
+    Notes
+    -----
+    The spectrum will need to have the uncertainty defined in order
+    for the SNR to be calculated.
+
+    Two methods implemented:
+      1) SNR based on uncertainty
+          This will compute the mean of the flux / uncertainty.
+
+      2) SNR based on noise_region
```

Comment:
I don't think this function should implement the `noise_region` option.

I agree the functionality should exist, though.  So I think this function should only do option 1 but at the same time there should be a function `uncertainty_from_noise_region`.  That would take a region and yield a *new* spectrum with the uncertainty set by that region, which the user would then pass to this function to acheive what you've got here.  That would also be more performant for the list-of-regions case because the noise calculation need only be done once.  If you agree with this,  @brechmos-stsci, feel free to add that function here, or in a follow-on PR.

The argument being that we need to resist having a few functions that do a bunch of things if the attitude for the analysis tools is a "toolbox" - in a tool box you have lots of tools to do a specific job well, not a single swiss army knife (obviously this is not as clear-cut as that, but still).

Reply:
I would suggest that both "versions" fit well in the `snr()` method. The method `snr` does one thing, it calculates the S/N.  The methods to define which uncertainty to use is very well separated by `noise_region` being None or not.  

Is the `uncertainty_from_noise_region()` method that  creates a new spectrum going to be used a lot outside of `snr()`?  If not, then I think it makes more sense to have the `noise_region` parameter for `snr()`.

So, I think it is a simpler, cleaner implementation to have `noise_region` as a parameter. But, I will do what you guys wish as you might have other thoughts of how this all fits together.  Maybe ping someone else for their opinion on this topic specifically?

Reply:
The point isn't so much the implemenation - I agree it's straightforward and *could* make sense here.  It's that I think the typical user thinks of the uncertainty in a spectrum as being a part of the spectrum, *not* a thing you calculate as part of the operation of computing the S/N.  In fact, in a lot of UV/optical/IR spectra, you get incorrect results from choosing a noise region like this because the uncertainty varies *a lot* across the spectrum (sky lines are the most relevant thing here, but so is the contribution to the uncertainty from, say a line itself). So in my mind having the ``noise_region`` favors a *particular* way of computing the noise, which is *known* to be wrong for one of the most astronomical use cases.

> Is the uncertainty_from_noise_region() method that creates a new spectrum going to be used a lot outside of snr()?

Yes, I would say so - if you aren't just stopping at computing `snr` (which most users aren't), the uncertainty will also be used in lots of other ways - nearly every analysis task.  E.g., weighting for resampling or line-fitting, judging if the line fluxes measured from a flux measurement are significant, etc.  So if the user *does* need this particular algorithm, they'll *need* `uncertainty_from_noise_region` anyway for many of their other analysis tasks.